src/source/src/manager.rs

```diff
@@ -133,7 +129,124 @@ pub struct MemSourceManager {
 
 #[async_trait]
 impl SourceManager for MemSourceManager {
-    async fn create_source(&self, source_id: &TableId, info: StreamSourceInfo) -> Result<()> {
+    fn get_source(&self, table_id: &TableId) -> Result<SourceDesc> {
+        let sources = self.get_sources()?;
+        sources.get(table_id).cloned().ok_or_else(|| {
+            InternalError(format!("Get source table id not exists: {:?}", table_id)).into()
+        })
+    }
+
+    fn drop_source(&self, table_id: &TableId) -> Result<()> {
+        let mut sources = self.get_sources()?;
+        ensure!(
+            sources.contains_key(table_id),
+            "Source does not exist: {:?}",
+            table_id
+        );
+        sources.remove(table_id);
+        Ok(())
+    }
+
+    fn clear_sources(&self) -> Result<()> {
+        let mut sources = self.get_sources()?;
+        sources.clear();
+        Ok(())
+    }
+
+    fn metrics(&self) -> Arc<SourceMetrics> {
+        self.metrics.clone()
+    }
+
+    fn msg_buf_size(&self) -> usize {
+        self.connector_message_buffer_size
+    }
+
+    fn get_sources(&self) -> Result<MutexGuard<'_, HashMap<TableId, SourceDesc>>> {
+        Ok(self.sources.lock())
+    }
+}
+
+impl Default for MemSourceManager {
+    fn default() -> Self {
+        MemSourceManager {
+            sources: Default::default(),
+            metrics: Default::default(),
+            connector_message_buffer_size: 16,
+        }
+    }
+}
+
+impl MemSourceManager {
+    pub fn new(metrics: Arc<SourceMetrics>, connector_message_buffer_size: usize) -> Self {
+        MemSourceManager {
+            sources: Mutex::new(HashMap::new()),
+            metrics,
+            connector_message_buffer_size,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct SourceDescBuilder {
+    id: TableId,
+    info: ProstSourceInfo,
+    mgr: SourceManagerRef,
+}
+
+impl SourceDescBuilder {
+    pub fn new(id: TableId, info: &ProstSourceInfo, mgr: &SourceManagerRef) -> Self {
+        Self {
+            id,
+            info: info.clone(),
+            mgr: mgr.clone(),
+        }
+    }
+
+    pub async fn build(&self) -> Result<SourceDesc> {
+        let Self { id, info, mgr } = self;
+        match &info {
+            ProstSourceInfo::TableSource(info) => Self::build_table_source(mgr, id, info),
+            ProstSourceInfo::StreamSource(info) => Self::build_stream_source(mgr, info).await,
+        }
+    }
+
+    fn build_table_source(
+        mgr: &SourceManagerRef,
+        table_id: &TableId,
+        info: &TableSourceInfo,
+    ) -> Result<SourceDesc> {
+        let mut sources = mgr.get_sources()?;
+        if let Some(source_desc) = sources.get(table_id) {
+            return Ok(source_desc.clone());
+        }
```

Comment:
I think it would be better to move this single-flight logic to the `TableSourceManager` (or something else) itself to avoid exposing the inner structure. Besides, the trait and the `Mem` impl can also be cleaned up.

Let's do this in future PRs!